Refactored linguist-backend

ahhhndre opened the pull request:
## Hey, I just made a Pull Request!

This PR includes several improvements to the Linguist backend:

- Added tests for the `LinguistBackendDatabase` and `LinguistBackendApi`
- Added support for using SQLite as a database
- Removed the default from the `processes_date` column
- Converted the `LinguistBackendApi` into an Interface
- Added the `LinguistBackendClient` which implements the `LinguistBackendApi` Interface
- Unprocessed entities will get processed before stale entities
- Entities in the Linguist database but not in the Catalog anymore will be deleted
- Improved the README's headings

Closes [#16394](https://github.com/backstage/backstage/issues/16394)

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

ahhhndre commented:
@jhaals some of the tests are failing and I could use some help with them. Testing it not one of my strengths but the only way I can improve is by writing them. My issue is around using the `CatalogClient` as you can see in the failed tests. Got any suggestions?

ahhhndre commented:
@elizabeth-a-hobbs @ThomasAwesome @JosiahCraw if you have some time and could give this a review I'd appreciate it. I'm not strong at testing so any tips to improve them would be welcome as well. Thanks in advance!

jhaals reviewed (changes requested):
Always nice to see tests being added!

I would double check what need to be public in the `LinguistBackendApi` as there is a lot of new methods being exposed now and some that do not take any arguments and do not return anything

ahhhndre commented on plugins/linguist-backend/api-report.md:
```diff
@@ -38,8 +40,18 @@ export class LinguistBackendApi {
     linguistJsOptions?: Record<string, unknown>,
   );
   // (undocumented)
+  addNewEntities(): Promise<void>;
```
Yes, I had to make these public to be able to test things more easily. If there's a way around this then I'd be happy to head down that direction.

ahhhndre replied:
I've gone done the route of creating an interface and that's allowed me to not make these all public but documented as internal. Now, I'm wondering if this is considered a breaking change?

ahhhndre commented on plugins/linguist-backend/migrations/20230216_remove_processed_date_default.js:
```diff
@@ -0,0 +1,39 @@
+  // Sqlite does not support this raw SQL
+  if (!knex.client.config.client.includes('sqlite3')) {
+    await knex.raw(
+      'ALTER TABLE entity_result ALTER COLUMN processed_date DROP DEFAULT;',
```
So the background here is that the `processed_date` column had a default that would set the current date automatically when a new record was added. That's actually not correct as entities get added to the table and then they get processed in batches. This change fixes this by removing the default.

ahhhndre commented:
> > @jhaals some of the tests are failing and I could use some help with them. Testing it not one of my strengths but the only way I can improve is by writing them. My issue is around using the `CatalogClient` as you can see in the failed tests. Got any suggestions?
> 
> I did a code search for `jest.mock('@backstage/catalog-client` and it looks like the other cases do not mock it inside the describe block so try moving it outside :)

Yes, I gave that a try but was also running into issue. I moved to passing this in from the router instead and also using the `CatalogApi`. That worked much better and helped me discover the actual issue was how I was doing the `entities.forEach()` which I changed to a `for of` loop instead

jhaals commented on plugins/linguist-backend/src/api/index.ts:
```diff
@@ -14,4 +14,5 @@
  */
 
-export { LinguistBackendApi } from './LinguistBackendApi';
+export { LinguistBackendClient } from './LinguistBackendClient';
```
You should be able to remove this export right? I'm thinking this is the default client that you get when you don't supply your own :)

ahhhndre replied:
Yes, good call. I've removed this 👍

ahhhndre commented:
Alright ready for another look @jhaals :)

jhaals commented on plugins/linguist-backend/api-report.md:
```diff
@@ -62,12 +54,16 @@ export class LinguistBackendDatabase implements LinguistBackendStore {
 
 // @public (undocumented)
 export interface LinguistBackendStore {
+  // (undocumented)
```
Isn't an `LinguistBackendStore` internal concern for the backend given that users only pass in the API(client)? Perhaps it could be unexported as well?

ahhhndre replied:
Yes, that does not need to be exported. I've removed it 👍

jhaals commented on plugins/linguist-backend/api-report.md:
```diff
@@ -49,11 +37,15 @@ export class LinguistBackendDatabase implements LinguistBackendStore {
+  getProcessedEntities(): Promise<ProcessedEntity[]>;
   // (undocumented)
-  getUnprocessedEntities(): Promise<string[] | []>;
+  getUnprocessedEntities(): Promise<string[]>;
```
Do these really need to be implemented by everyone? It looks like a lot of these are for internal workings and testings.

I would try to slim down the interface to only what's needed for the backend to function and construct as much as possibly inside the router(privately) unless there really need to be possibly to supply alternatives.
Your default client that's not exported could have all of these "public" but internal for you to properly test if that makes sense :)

ahhhndre replied:
This should all be resolved as well 👍

jhaals commented:
I took another look @ahhhndre and I hope you don't feel that I am blocking you from getting this in 😅

ahhhndre commented:
> I took another look @ahhhndre and I hope you don't feel that I am blocking you from getting this in 😅

No worries, this is helping me make my design better so I'm happy to get the feedback. Many thanks @jhaals and this is ready for another look 🚀

jhaals commented on plugins/linguist-backend/src/api/LinguistBackendClient.ts:
```diff
@@ -148,15 +170,17 @@ export class LinguistBackendApi {
       try {
         await this.generateEntityLanguages(entityRef, url);
       } catch (error) {
+        console.log(error);
```
This probably snuck in :)

ahhhndre replied:
Yes, good catch. It's gone now 👍

jhaals commented on plugins/linguist-backend/src/db/LinguistBackendDatabase.ts:
```diff
@@ -102,13 +104,13 @@ export class LinguistBackendDatabase implements LinguistBackendStore {
 
     try {
-      return JSON.parse(entityResults.languages);
+      return JSON.parse(entityResults.languages as string);
```
is this cast really needed inside JSON.parse?

ahhhndre replied:
Yeah, this can be refactored. Thanks!

jhaals commented on plugins/linguist-backend/src/db/LinguistBackendDatabase.ts:
```diff
@@ -118,9 +120,20 @@ export class LinguistBackendDatabase implements LinguistBackendStore {
       const processEntity = {
         entityRef: rawEntity.entity_ref,
-        processedDate: rawEntity.processed_date,
+        processedDate: processedDate,
```
```suggestion
        processedDate,
```

ahhhndre replied:
Done 👍

jhaals reviewed (changes requested):
Looking a lot better 👏 

The only important thing is to update the changeset accordingly by looking at the public api in the api-report.

So what I'm thinking that this needs to be a `minor` change mentioning 

```md
**BREAKING**: Removed public constructor from `LinguistBackendApi`. Removed export of `LinguistBackendDatabase` and `LinguistBackendDatabase`
```

ahhhndre commented:
> Looking a lot better 👏
> 
> The only important thing is to update the changeset accordingly by looking at the public api in the api-report.
> 
> So what I'm thinking that this needs to be a `minor` change mentioning
> 
> ```lisp
> **BREAKING**: Removed public constructor from `LinguistBackendApi`. Removed export of `LinguistBackendDatabase` and `LinguistBackendDatabase`
> ```

Alright, I've resolved the latest comments and updated the changeset with the information about the breaking changes and set it to `minor`. Ready for another look when you can @jhaals 🚀

jhaals commented on plugins/linguist-backend/src/service/router.ts:
```diff
@@ -74,14 +76,16 @@ export async function createRouter(
+  const linguistBackendClient =
     routerOptions.linguistBackendApi ||
-    new LinguistBackendApi(
+    new LinguistBackendClient(
```
I think we can avoid creating the `linguistBackendStore` entierly if the user passes in a `linguistBackendApi`.

Could it be perhaps it worth dropping backend from the name? So users can pass in their own `LinguistApi` instead of `LinguistBackendApi`?

ahhhndre replied:
Sorry, finally got around to this. I'm very confused, I'm not sure what you mean? If I removed `linguistBackendStore` I still need to create it somewhere if I did that in the `linguistBackendApi` I'd still need to bass it the `database` (`DatabaseManager`), right? I feel like I'm missing something obvious.

As fro the name their is already a `LinguistApi` in the frontend so would like to keep this name to make things more clear if that's alright?

jhaals replied:
On line 75 we create the `linguistBackendStore` but we only use it if the user does not pass in their own `linguistBackendApi` so I'm thinking we can avoid creating it unless it is needed. You're right that we still have to pass in the databaseManager but it wouldn't do any work unless needed.

jhaals reviewed (approved):
Consider my last comment an optional optimisation @ahhhndre. I don't want to block your work on this as there's some bug fixes in here as well. Feel free to merge this in when you feel ready.

jhaals commented:
Actually I'm going to merge this in now, feel like it's time :D

awanlin commented:
Thanks a ton for taking the time to review this and help me improve the architecture of this plugin @jhaals! 🚀
